### PR TITLE
Add decomposition for CCZ gate and IonQTargetGateset when qubits are all-to-all connected

### PR DESCRIPTION
**Pull request by yinghui-hu**
When the 3 qubits are all connected to each other, the decomposition can be further optimized https://github.com/quantumlib/Cirq/issues/6068. Next step is to use all-to-all connectivity when decomposing circuits using `IonQTargetGateset`.

### CODE REVIEW

**Comment by tanujkhattar** [comment]
Please see my comment here - https://github.com/quantumlib/Cirq/issues/6068#issuecomment-1557615265

Adding an alternative decomposition as done here is not enough because you would need a way to propagate the `all_to_all_connect` flag through `cirq.decompose` which is currently not supported.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Please see my comment here - https://github.com/quantumlib/Cirq/issues/6068#issuecomment-1557615265

Adding an alternative decomposition as done here is not enough because you would need a way to propagate the `all_to_all_connect` flag through `cirq.decompose` which is currently not supported.

**Comment by yinghui-hu** [comment]
> 

Done. Thanks for the guidance.

**Comment by yinghui-hu** [comment]
> 

Requested changes done. Please review again. Thanks. @tanujkhattar

**Comment by yinghui-hu** [comment]
> 

Done

File: cirq-core/cirq/ops/three_qubit_gates.py [hunk 1]
```diff
@@ -217,6 +217,55 @@ def controlled(
         )
 
 
+def decompose_all_to_all_connect_ccz_gate(
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/three_qubit_gates.py:220]: Move this decomposition and corresponding test to `cirq-ionq/cirq_ionq/ionq_gateset.py`.

File: cirq-core/cirq/ops/three_qubit_gates.py [hunk 2]
```diff
@@ -217,6 +217,55 @@ def controlled(
...
+    ccz_gate: 'CCZPowGate', qubits: Tuple['cirq.Qid', ...]
+) -> 'cirq.OP_TREE':
+    """If qubits are all-to-all connected, e.g. qubits in the same ion trap,
+    the decomposition will be:
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/three_qubit_gates.py:224]: Summary line in the docstring should be a single line with <= 100 characters followed by an empty newline and then a detailed description if needed. Please update here and elsewhere

File: cirq-ionq/cirq_ionq/ionq_gateset.py [hunk 3]
```diff
@@ -78,12 +78,20 @@ def _decompose_two_qubit_operation(self, op: cirq.Operation, _) -> cirq.OP_TREE:
...
 
+    def _decompose_multi_qubit_operation(self, op: cirq.Operation, _) -> cirq.OP_TREE:
+        if isinstance(op.gate, cirq.CCZPowGate):
+            return cirq.ops.decompose_all_to_all_connect_ccz_gate(op.gate, op.qubits)
+        return NotImplemented
+
     @property
     def preprocess_transformers(self) -> List['cirq.TRANSFORMER']:
-        """List of transformers which should be run before decomposing individual operations."""
+        """List of transformers which should be run before decomposing individual operations.
```
Comment (tanujkhattar) [inline cirq-ionq/cirq_ionq/ionq_gateset.py:88]: Add a new line after summary line and before description. 
```suggestion
        """List of transformers which should be run before decomposing individual operations.
       
```

File: cirq-core/cirq/ops/__init__.py [hunk 4]
```diff
@@ -201,6 +201,7 @@
     CCXPowGate,
     CCZ,
     CCZPowGate,
+    decompose_all_to_all_connect_ccz_gate,
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/__init__.py:204]: Remove from `cirq/ops/__init__.py`

**Review by tanujkhattar — APPROVED** [review]
LGTM % some final comments

File: cirq-core/cirq/ops/three_qubit_gates.py (hunk 1 above)
Comment (yinghui-hu) [reply cirq-core/cirq/ops/three_qubit_gates.py]: Done

File: cirq-core/cirq/ops/three_qubit_gates.py (hunk 2 above)
Comment (yinghui-hu) [reply cirq-core/cirq/ops/three_qubit_gates.py]: Done. Thanks for the guide.

File: cirq-core/cirq/ops/__init__.py (hunk 4 above)
Comment (yinghui-hu) [reply cirq-core/cirq/ops/__init__.py]: Done

**Comment by yinghui-hu** [comment]
This PR is ready to merge @tanujkhattar Thanks!

**Pull request merged**